File: javascript/src/model/CallLogInfo.js
(function(root, factory) {
  if (typeof define === 'function' && define.amd) {
    // AMD. Register as an anonymous module.
    define(['../ApiClient', './CallerInfo', './RecordingInfo'], factory);
  } else if (typeof module === 'object' && module.exports) {
    // CommonJS-like environments that support module.exports, like Node.
    module.exports = factory(require('../ApiClient'), require('./CallerInfo'), require('./RecordingInfo'));
  } else {
    // Browser globals (root is window)
    if (!root.RingCentralApi) {
      root.RingCentralApi = {};
    }
    root.RingCentralApi.CallLogInfo = factory(root.RingCentralApi.ApiClient, root.RingCentralApi.CallerInfo, root.RingCentralApi.RecordingInfo);
  }
}(this, function(ApiClient, CallerInfo, RecordingInfo) {
  'use strict';

  /**
   * The CallLogInfo model module.
   * @module model/CallLogInfo
   * @version v1.0
   */

  /**
   * Constructs a new <code>CallLogInfo</code>.
   * @alias module:model/CallLogInfo
   * @class
   */
  var exports = function() {













  };

  /**
   * Constructs a <code>CallLogInfo</code> from a plain JavaScript object, optionally creating a new instance.
   * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
   * @param {Object} data The plain JavaScript object bearing properties of interest.
   * @param {module:model/CallLogInfo} obj Optional instance to populate.
   * @return {module:model/CallLogInfo} The populated <code>CallLogInfo</code> instance.
   */
  exports.constructFromObject = function(data, obj) {
    if (data) { 
      obj = obj || new exports();

      if (data.hasOwnProperty('id')) {
        obj['id'] = ApiClient.convertToType(data['id'], 'String');
      }
      if (data.hasOwnProperty('uri')) {
        obj['uri'] = ApiClient.convertToType(data['uri'], 'String');
      }
      if (data.hasOwnProperty('sessionId')) {
        obj['sessionId'] = ApiClient.convertToType(data['sessionId'], 'String');
      }
      if (data.hasOwnProperty('from')) {
        obj['from'] = CallerInfo.constructFromObject(data['from']);
      }
      if (data.hasOwnProperty('to')) {
        obj['to'] = CallerInfo.constructFromObject(data['to']);
      }
      if (data.hasOwnProperty('type')) {
        obj['type'] = ApiClient.convertToType(data['type'], 'String');
      }
      if (data.hasOwnProperty('direction')) {
        obj['direction'] = ApiClient.convertToType(data['direction'], 'String');
      }
      if (data.hasOwnProperty('action')) {
        obj['action'] = ApiClient.convertToType(data['action'], 'String');
      }
      if (data.hasOwnProperty('result')) {
        obj['result'] = ApiClient.convertToType(data['result'], 'String');
      }
      if (data.hasOwnProperty('startTime')) {
        obj['startTime'] = ApiClient.convertToType(data['startTime'], 'Date');
      }
      if (data.hasOwnProperty('duration')) {
        obj['duration'] = ApiClient.convertToType(data['duration'], 'Integer');
      }
      if (data.hasOwnProperty('recording')) {
        obj['recording'] = RecordingInfo.constructFromObject(data['recording']);
      }
    }
    return obj;
  }


  /**
   * Internal identifier of a cal log record
   * @member {String} id
   */
  exports.prototype['id'] = undefined;

  /**
   * Canonical URI of a call log record
   * @member {String} uri
   */
  exports.prototype['uri'] = undefined;

  /**
   * Internal identifier of a call session
   * @member {String} sessionId
   */
  exports.prototype['sessionId'] = undefined;

  /**
   * @member {module:model/CallerInfo} from
   */
  exports.prototype['from'] = undefined;

  /**
   * @member {module:model/CallerInfo} to
   */
  exports.prototype['to'] = undefined;

  /**
   * Call type
   * @member {module:model/CallLogInfo.TypeEnum} type
   */
  exports.prototype['type'] = undefined;

  /**
   * Call direction
   * @member {module:model/CallLogInfo.DirectionEnum} direction
   */
  exports.prototype['direction'] = undefined;

  /**
   * Action description of the call operation
   * @member {module:model/CallLogInfo.ActionEnum} action
   */
  exports.prototype['action'] = undefined;

  /**
   * Status description of the call operation
   * @member {module:model/CallLogInfo.ResultEnum} result
   */
  exports.prototype['result'] = undefined;

  /**
   * The call start datetime in ISO 8601 format including timezone, for example 2016-03-10T18:07:52.534Z
   * @member {Date} startTime
   */
  exports.prototype['startTime'] = undefined;

  /**
   * Call duration in seconds
   * @member {Integer} duration
   */
  exports.prototype['duration'] = undefined;

  /**
   * @member {module:model/RecordingInfo} recording
   */
  exports.prototype['recording'] = undefined;


  /**
   * Allowed values for the <code>type</code> property.
   * @enum {String}
   * @readonly
   */
  exports.TypeEnum = { 
    /**
     * value: Voice
     * @const
     */
    VOICE: "Voice",
    
    /**
     * value: Fax
     * @const
     */
    FAX: "Fax"
  };  /**
   * Allowed values for the <code>direction</code> property.
   * @enum {String}
   * @readonly
   */
  exports.DirectionEnum = { 
    /**
     * value: Inbound
     * @const
     */
    INBOUND: "Inbound",
    
    /**
     * value: Outbound
     * @const
     */
    OUTBOUND: "Outbound"
  };  /**
   * Allowed values for the <code>action</code> property.
   * @enum {String}
   * @readonly
   */
  exports.ActionEnum = { 
    /**
     * value: Unknown
     * @const
     */
    UNKNOWN: "Unknown",
    
    /**
     * value: Phone Call
     * @const
     */
    PHONE_CALL: "Phone Call",
    
    /**
     * value: Phone Login
     * @const
     */
    PHONE_LOGIN: "Phone Login",
    
    /**
     * value: Incoming Fax
     * @const
     */
    INCOMING_FAX: "Incoming Fax",
    
    /**
     * value: Accept Call
     * @const
     */
    ACCEPT_CALL: "Accept Call",
    
    /**
     * value: FindMe
     * @const
     */
    FINDME: "FindMe",
    
    /**
     * value: FollowMe
     * @const
     */
    FOLLOWME: "FollowMe",
    
    /**
     * value: Outgoing Fax
     * @const
     */
    OUTGOING_FAX: "Outgoing Fax",
    
    /**
     * value: Call Return
     * @const
     */
    CALL_RETURN: "Call Return",
    
    /**
     * value: Calling Card
     * @const
     */
    CALLING_CARD: "Calling Card",
    
    /**
     * value: Ring Directly
     * @const
     */
    RING_DIRECTLY: "Ring Directly",
    
    /**
     * value: RingOut Web
     * @const
     */
    RINGOUT_WEB: "RingOut Web",
    
    /**
     * value: VoIP Call
     * @const
     */
    VOIP_CALL: "VoIP Call",
    
    /**
     * value: RingOut PC
     * @const
     */
    RINGOUT_PC: "RingOut PC",
    
    /**
     * value: RingMe
     * @const
     */
    RINGME: "RingMe",
    
    /**
     * value: Transfer
     * @const
     */
    TRANSFER: "Transfer",
    
    /**
     * value: 411 Info
     * @const
     */
    _411_INFO: "411 Info",
    
    /**
     * value: Emergency
     * @const
     */
    EMERGENCY: "Emergency",
    
    /**
     * value: E911 Update
     * @const
     */
    E911_UPDATE: "E911 Update",
    
    /**
     * value: Support
     * @const
     */
    SUPPORT: "Support",
    
    /**
     * value: RingOut Mobile
     * @const
     */
    RINGOUT_MOBILE: "RingOut Mobile"
  };  /**
   * Allowed values for the <code>result</code> property.
   * @enum {String}
   * @readonly
   */
  exports.ResultEnum = { 
    /**
     * value: Unknown
     * @const
     */
    UNKNOWN: "Unknown",
    
    /**
     * value: ResultInProgress
     * @const
     */
    RESULTINPROGRESS: "ResultInProgress",
    
    /**
     * value: Missed
     * @const
     */
    MISSED: "Missed",
    
    /**
     * value: Call accepted
     * @const
     */
    CALL_ACCEPTED: "Call accepted",
    
    /**
     * value: Voicemail
     * @const
     */
    VOICEMAIL: "Voicemail",
    
    /**
     * value: Rejected
     * @const
     */
    REJECTED: "Rejected",
    
    /**
     * value: Reply
     * @const
     */
    REPLY: "Reply",
    
    /**
     * value: Received
     * @const
     */
    RECEIVED: "Received",
    
    /**
     * value: Receive Error
     * @const
     */
    RECEIVE_ERROR: "Receive Error",
    
    /**
     * value: Fax on Demand
     * @const
     */
    FAX_ON_DEMAND: "Fax on Demand",
    
    /**
     * value: Partial Receive
     * @const
     */
    PARTIAL_RECEIVE: "Partial Receive",
    
    /**
     * value: Blocked
     * @const
     */
    BLOCKED: "Blocked",
    
    /**
     * value: Call connected
     * @const
     */
    CALL_CONNECTED: "Call connected",
    
    /**
     * value: No Answer
     * @const
     */
    NO_ANSWER: "No Answer",
    
    /**
     * value: International Disabled
     * @const
     */
    INTERNATIONAL_DISABLED: "International Disabled",
    
    /**
     * value: Busy
     * @const
     */
    BUSY: "Busy",
    
    /**
     * value: Send Error
     * @const
     */
    SEND_ERROR: "Send Error",
    
    /**
     * value: Sent
     * @const
     */
    SENT: "Sent",
    
    /**
     * value: No fax machine
     * @const
     */
    NO_FAX_MACHINE: "No fax machine",
    
    /**
     * value: ResultEmpty
     * @const
     */
    RESULTEMPTY: "ResultEmpty",
    
    /**
     * value: Account
     * @const
     */
    ACCOUNT: "Account",
    
    /**
     * value: Suspended
     * @const
     */
    SUSPENDED: "Suspended",
    
    /**
     * value: Call Failed
     * @const
     */
    CALL_FAILED: "Call Failed",
    
    /**
     * value: Call Failure
     * @const
     */
    CALL_FAILURE: "Call Failure",
    
    /**
     * value: Internal Error
     * @const
     */
    INTERNAL_ERROR: "Internal Error",
    
    /**
     * value: IP Phone offline
     * @const
     */
    IP_PHONE_OFFLINE: "IP Phone offline",
    
    /**
     * value: Restricted Number
     * @const
     */
    RESTRICTED_NUMBER: "Restricted Number",
    
    /**
     * value: Wrong Number
     * @const
     */
    WRONG_NUMBER: "Wrong Number",
    
    /**
     * value: Stopped
     * @const
     */
    STOPPED: "Stopped",
    
    /**
     * value: Hang up
     * @const
     */
    HANG_UP: "Hang up",
    
    /**
     * value: Poor Line Quality
     * @const
     */
    POOR_LINE_QUALITY: "Poor Line Quality",
    
    /**
     * value: Partially Sent
     * @const
     */
    PARTIALLY_SENT: "Partially Sent",
    
    /**
     * value: International Restriction
     * @const
     */
    INTERNATIONAL_RESTRICTION: "International Restriction",
    
    /**
     * value: Abandoned
     * @const
     */
    ABANDONED: "Abandoned",
    
    /**
     * value: Declined
     * @const
     */
    DECLINED: "Declined",
    
    /**
     * value: Fax Receipt Error
     * @const
     */
    FAX_RECEIPT_ERROR: "Fax Receipt Error",
    
    /**
     * value: Fax Send Error
     * @const
     */
    FAX_SEND_ERROR: "Fax Send Error"
  };

  return exports;
}));

File: csharp/src/main/csharp/IO/Swagger/Model/CallLogInfo.cs
using System;
using System.Linq;
using System.IO;
using System.Text;
using System.Collections;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Runtime.Serialization;
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;

namespace IO.Swagger.Model
{
    /// <summary>
    /// 
    /// </summary>
    [DataContract]
    public partial class CallLogInfo :  IEquatable<CallLogInfo>
    { 
    
        /// <summary>
        /// Call type
        /// </summary>
        /// <value>Call type</value>
        [JsonConverter(typeof(StringEnumConverter))]
        public enum TypeEnum {
            
            [EnumMember(Value = "Voice")]
            Voice,
            
            [EnumMember(Value = "Fax")]
            Fax
        }

    
        /// <summary>
        /// Call direction
        /// </summary>
        /// <value>Call direction</value>
        [JsonConverter(typeof(StringEnumConverter))]
        public enum DirectionEnum {
            
            [EnumMember(Value = "Inbound")]
            Inbound,
            
            [EnumMember(Value = "Outbound")]
            Outbound
        }

    
        /// <summary>
        /// Action description of the call operation
        /// </summary>
        /// <value>Action description of the call operation</value>
        [JsonConverter(typeof(StringEnumConverter))]
        public enum ActionEnum {
            
            [EnumMember(Value = "Unknown")]
            Unknown,
            
            [EnumMember(Value = "Phone Call")]
            PhoneCall,
            
            [EnumMember(Value = "Phone Login")]
            PhoneLogin,
            
            [EnumMember(Value = "Incoming Fax")]
            IncomingFax,
            
            [EnumMember(Value = "Accept Call")]
            AcceptCall,
            
            [EnumMember(Value = "FindMe")]
            Findme,
            
            [EnumMember(Value = "FollowMe")]
            Followme,
            
            [EnumMember(Value = "Outgoing Fax")]
            OutgoingFax,
            
            [EnumMember(Value = "Call Return")]
            CallReturn,
            
            [EnumMember(Value = "Calling Card")]
            CallingCard,
            
            [EnumMember(Value = "Ring Directly")]
            RingDirectly,
            
            [EnumMember(Value = "RingOut Web")]
            RingoutWeb,
            
            [EnumMember(Value = "VoIP Call")]
            VoipCall,
            
            [EnumMember(Value = "RingOut PC")]
            RingoutPc,
            
            [EnumMember(Value = "RingMe")]
            Ringme,
            
            [EnumMember(Value = "Transfer")]
            Transfer,
            
            [EnumMember(Value = "411 Info")]
            _411Info,
            
            [EnumMember(Value = "Emergency")]
            Emergency,
            
            [EnumMember(Value = "E911 Update")]
            E911Update,
            
            [EnumMember(Value = "Support")]
            Support,
            
            [EnumMember(Value = "RingOut Mobile")]
            RingoutMobile
        }

    
        /// <summary>
        /// Status description of the call operation
        /// </summary>
        /// <value>Status description of the call operation</value>
        [JsonConverter(typeof(StringEnumConverter))]
        public enum ResultEnum {
            
            [EnumMember(Value = "Unknown")]
            Unknown,
            
            [EnumMember(Value = "ResultInProgress")]
            Resultinprogress,
            
            [EnumMember(Value = "Missed")]
            Missed,
            
            [EnumMember(Value = "Call accepted")]
            CallAccepted,
            
            [EnumMember(Value = "Voicemail")]
            Voicemail,
            
            [EnumMember(Value = "Rejected")]
            Rejected,
            
            [EnumMember(Value = "Reply")]
            Reply,
            
            [EnumMember(Value = "Received")]
            Received,
            
            [EnumMember(Value = "Receive Error")]
            ReceiveError,
            
            [EnumMember(Value = "Fax on Demand")]
            FaxOnDemand,
            
            [EnumMember(Value = "Partial Receive")]
            PartialReceive,
            
            [EnumMember(Value = "Blocked")]
            Blocked,
            
            [EnumMember(Value = "Call connected")]
            CallConnected,
            
            [EnumMember(Value = "No Answer")]
            NoAnswer,
            
            [EnumMember(Value = "International Disabled")]
            InternationalDisabled,
            
            [EnumMember(Value = "Busy")]
            Busy,
            
            [EnumMember(Value = "Send Error")]
            SendError,
            
            [EnumMember(Value = "Sent")]
            Sent,
            
            [EnumMember(Value = "No fax machine")]
            NoFaxMachine,
            
            [EnumMember(Value = "ResultEmpty")]
            Resultempty,
            
            [EnumMember(Value = "Account")]
            Account,
            
            [EnumMember(Value = "Suspended")]
            Suspended,
            
            [EnumMember(Value = "Call Failed")]
            CallFailed,
            
            [EnumMember(Value = "Call Failure")]
            CallFailure,
            
            [EnumMember(Value = "Internal Error")]
            InternalError,
            
            [EnumMember(Value = "IP Phone offline")]
            IpPhoneOffline,
            
            [EnumMember(Value = "Restricted Number")]
            RestrictedNumber,
            
            [EnumMember(Value = "Wrong Number")]
            WrongNumber,
            
            [EnumMember(Value = "Stopped")]
            Stopped,
            
            [EnumMember(Value = "Hang up")]
            HangUp,
            
            [EnumMember(Value = "Poor Line Quality")]
            PoorLineQuality,
            
            [EnumMember(Value = "Partially Sent")]
            PartiallySent,
            
            [EnumMember(Value = "International Restriction")]
            InternationalRestriction,
            
            [EnumMember(Value = "Abandoned")]
            Abandoned,
            
            [EnumMember(Value = "Declined")]
            Declined,
            
            [EnumMember(Value = "Fax Receipt Error")]
            FaxReceiptError,
            
            [EnumMember(Value = "Fax Send Error")]
            FaxSendError
        }

    
        /// <summary>
        /// Call type
        /// </summary>
        /// <value>Call type</value>
        [DataMember(Name="type", EmitDefaultValue=false)]
        public TypeEnum? Type { get; set; }
    
        /// <summary>
        /// Call direction
        /// </summary>
        /// <value>Call direction</value>
        [DataMember(Name="direction", EmitDefaultValue=false)]
        public DirectionEnum? Direction { get; set; }
    
        /// <summary>
        /// Action description of the call operation
        /// </summary>
        /// <value>Action description of the call operation</value>
        [DataMember(Name="action", EmitDefaultValue=false)]
        public ActionEnum? Action { get; set; }
    
        /// <summary>
        /// Status description of the call operation
        /// </summary>
        /// <value>Status description of the call operation</value>
        [DataMember(Name="result", EmitDefaultValue=false)]
        public ResultEnum? Result { get; set; }
    
        /// <summary>
        /// Initializes a new instance of the <see cref="CallLogInfo" /> class.
        /// Initializes a new instance of the <see cref="CallLogInfo" />class.
        /// </summary>
        /// <param name="Id">Internal identifier of a cal log record.</param>
        /// <param name="Uri">Canonical URI of a call log record.</param>
        /// <param name="SessionId">Internal identifier of a call session.</param>
        /// <param name="From">From.</param>
        /// <param name="To">To.</param>
        /// <param name="Type">Call type.</param>
        /// <param name="Direction">Call direction.</param>
        /// <param name="Action">Action description of the call operation.</param>
        /// <param name="Result">Status description of the call operation.</param>
        /// <param name="StartTime">The call start datetime in ISO 8601 format including timezone, for example 2016-03-10T18:07:52.534Z.</param>
        /// <param name="Duration">Call duration in seconds.</param>
        /// <param name="Recording">Recording.</param>

        public CallLogInfo(string Id = null, string Uri = null, string SessionId = null, CallerInfo From = null, CallerInfo To = null, TypeEnum? Type = null, DirectionEnum? Direction = null, ActionEnum? Action = null, ResultEnum? Result = null, DateTime? StartTime = null, int? Duration = null, RecordingInfo Recording = null)
        {
            this.Id = Id;
            this.Uri = Uri;
            this.SessionId = SessionId;
            this.From = From;
            this.To = To;
            this.Type = Type;
            this.Direction = Direction;
            this.Action = Action;
            this.Result = Result;
            this.StartTime = StartTime;
            this.Duration = Duration;
            this.Recording = Recording;
            
        }
        
    
        /// <summary>
        /// Internal identifier of a cal log record
        /// </summary>
        /// <value>Internal identifier of a cal log record</value>
        [DataMember(Name="id", EmitDefaultValue=false)]
        public string Id { get; set; }
    
        /// <summary>
        /// Canonical URI of a call log record
        /// </summary>
        /// <value>Canonical URI of a call log record</value>
        [DataMember(Name="uri", EmitDefaultValue=false)]
        public string Uri { get; set; }
    
        /// <summary>
        /// Internal identifier of a call session
        /// </summary>
        /// <value>Internal identifier of a call session</value>
        [DataMember(Name="sessionId", EmitDefaultValue=false)]
        public string SessionId { get; set; }
    
        /// <summary>
        /// Gets or Sets From
        /// </summary>
        [DataMember(Name="from", EmitDefaultValue=false)]
        public CallerInfo From { get; set; }
    
        /// <summary>
        /// Gets or Sets To
        /// </summary>
        [DataMember(Name="to", EmitDefaultValue=false)]
        public CallerInfo To { get; set; }
    
        /// <summary>
        /// The call start datetime in ISO 8601 format including timezone, for example 2016-03-10T18:07:52.534Z
        /// </summary>
        /// <value>The call start datetime in ISO 8601 format including timezone, for example 2016-03-10T18:07:52.534Z</value>
        [DataMember(Name="startTime", EmitDefaultValue=false)]
        public DateTime? StartTime { get; set; }
    
        /// <summary>
        /// Call duration in seconds
        /// </summary>
        /// <value>Call duration in seconds</value>
        [DataMember(Name="duration", EmitDefaultValue=false)]
        public int? Duration { get; set; }
    
        /// <summary>
        /// Gets or Sets Recording
        /// </summary>
        [DataMember(Name="recording", EmitDefaultValue=false)]
        public RecordingInfo Recording { get; set; }
    
        /// <summary>
        /// Returns the string presentation of the object
        /// </summary>
        /// <returns>String presentation of the object</returns>
        public override string ToString()
        {
            var sb = new StringBuilder();
            sb.Append("class CallLogInfo {\n");
            sb.Append("  Id: ").Append(Id).Append("\n");
            sb.Append("  Uri: ").Append(Uri).Append("\n");
            sb.Append("  SessionId: ").Append(SessionId).Append("\n");
            sb.Append("  From: ").Append(From).Append("\n");
            sb.Append("  To: ").Append(To).Append("\n");
            sb.Append("  Type: ").Append(Type).Append("\n");
            sb.Append("  Direction: ").Append(Direction).Append("\n");
            sb.Append("  Action: ").Append(Action).Append("\n");
            sb.Append("  Result: ").Append(Result).Append("\n");
            sb.Append("  StartTime: ").Append(StartTime).Append("\n");
            sb.Append("  Duration: ").Append(Duration).Append("\n");
            sb.Append("  Recording: ").Append(Recording).Append("\n");
            
            sb.Append("}\n");
            return sb.ToString();
        }
  
        /// <summary>
        /// Returns the JSON string presentation of the object
        /// </summary>
        /// <returns>JSON string presentation of the object</returns>
        public string ToJson()
        {
            return JsonConvert.SerializeObject(this, Formatting.Indented);
        }

        /// <summary>
        /// Returns true if objects are equal
        /// </summary>
        /// <param name="obj">Object to be compared</param>
        /// <returns>Boolean</returns>
        public override bool Equals(object obj)
        {
            // credit: http://stackoverflow.com/a/10454552/677735
            return this.Equals(obj as CallLogInfo);
        }

        /// <summary>
        /// Returns true if CallLogInfo instances are equal
        /// </summary>
        /// <param name="other">Instance of CallLogInfo to be compared</param>
        /// <returns>Boolean</returns>
        public bool Equals(CallLogInfo other)
        {
            // credit: http://stackoverflow.com/a/10454552/677735
            if (other == null)
                return false;

            return 
                (
                    this.Id == other.Id ||
                    this.Id != null &&
                    this.Id.Equals(other.Id)
                ) && 
                (
                    this.Uri == other.Uri ||
                    this.Uri != null &&
                    this.Uri.Equals(other.Uri)
                ) && 
                (
                    this.SessionId == other.SessionId ||
                    this.SessionId != null &&
                    this.SessionId.Equals(other.SessionId)
                ) && 
                (
                    this.From == other.From ||
                    this.From != null &&
                    this.From.Equals(other.From)
                ) && 
                (
                    this.To == other.To ||
                    this.To != null &&
                    this.To.Equals(other.To)
                ) && 
                (
                    this.Type == other.Type ||
                    this.Type != null &&
                    this.Type.Equals(other.Type)
                ) && 
                (
                    this.Direction == other.Direction ||
                    this.Direction != null &&
                    this.Direction.Equals(other.Direction)
                ) && 
                (
                    this.Action == other.Action ||
                    this.Action != null &&
                    this.Action.Equals(other.Action)
                ) && 
                (
                    this.Result == other.Result ||
                    this.Result != null &&
                    this.Result.Equals(other.Result)
                ) && 
                (
                    this.StartTime == other.StartTime ||
                    this.StartTime != null &&
                    this.StartTime.Equals(other.StartTime)
                ) && 
                (
                    this.Duration == other.Duration ||
                    this.Duration != null &&
                    this.Duration.Equals(other.Duration)
                ) && 
                (
                    this.Recording == other.Recording ||
                    this.Recording != null &&
                    this.Recording.Equals(other.Recording)
                );
        }

        /// <summary>
        /// Gets the hash code
        /// </summary>
        /// <returns>Hash code</returns>
        public override int GetHashCode()
        {
            // credit: http://stackoverflow.com/a/263416/677735
            unchecked // Overflow is fine, just wrap
            {
                int hash = 41;
                // Suitable nullity checks etc, of course :)
                
                if (this.Id != null)
                    hash = hash * 59 + this.Id.GetHashCode();
                
                if (this.Uri != null)
                    hash = hash * 59 + this.Uri.GetHashCode();
                
                if (this.SessionId != null)
                    hash = hash * 59 + this.SessionId.GetHashCode();
                
                if (this.From != null)
                    hash = hash * 59 + this.From.GetHashCode();
                
                if (this.To != null)
                    hash = hash * 59 + this.To.GetHashCode();
                
                if (this.Type != null)
                    hash = hash * 59 + this.Type.GetHashCode();
                
                if (this.Direction != null)
                    hash = hash * 59 + this.Direction.GetHashCode();
                
                if (this.Action != null)
                    hash = hash * 59 + this.Action.GetHashCode();
                
                if (this.Result != null)
                    hash = hash * 59 + this.Result.GetHashCode();
                
                if (this.StartTime != null)
                    hash = hash * 59 + this.StartTime.GetHashCode();
                
                if (this.Duration != null)
                    hash = hash * 59 + this.Duration.GetHashCode();
                
                if (this.Recording != null)
                    hash = hash * 59 + this.Recording.GetHashCode();
                
                return hash;
            }
        }

    }
}
